### PR TITLE
ADM-573:[chore] refactor E2E codes and add hook for E2E

### PR DESCRIPTION
**Pull request by SimonTal**
## Summary

...

## Before

_Description_
There were some duplicated testing case before, or unreasonable waiting time, resulting in a longer test time.

**Screenshots**
before

## After

_Description_
Refactor E2E codes and add hook for E2E and change timeout config of Jest, Cypress for slower PC. 

- new npm command
`npm run e2e:local`
`npm run e2e:with-server`

**Screenshots**
after

## Note

_Null_

### CODE REVIEW

File: frontend/cypress/e2e/createANewProject.cy.ts [hunk 1]
```diff
@@ -246,48 +261,58 @@ describe('Create a new project', () => {
...
-    nextButton().should('be.disabled')
+    configPage.nextStepButton.should('be.disabled')
 
     configPage.selectMetricsData()
 
     configPage.fillBoardInfoAndVerifyWithClassicJira('1963', 'test@test.com', 'PLL', 'site', 'mockToken')
-
-    cy.contains('Verified').should('exist')
-    cy.contains('Reset').should('exist')
+    configPage.getVerifiedButton(configPage.boardConfigSection).should('be.disabled')
```
Comment (Zxx-Ivy) [inline frontend/cypress/e2e/createANewProject.cy.ts:269]: These two lines don't seem to line up.

File: frontend/scripts/runE2eWithServer.ts [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+import process from 'process'
+
+const PORT = {
+  FE: 4321,
+  BE: 4322,
+  STUB: 4323,
+}
+
+const DIR = {
+  FE: path.resolve(__dirname, '../'),
```
Comment (guzhongren) [inline frontend/scripts/runE2eWithServer.ts:12]: should not use short name, make it readable for everyone

File: frontend/scripts/runE2eWithServer.ts [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+}
+
+const main = async (args: string[]) => {
+  const { $ } = await import('execa')
+
+  const checkPorts = async () => {
+    const ports = Object.values(PORT)
+    const checkPorts = ports.map((port) => tcpPortUsed.check(port))
+
+    const usingPorts = await Promise.all(checkPorts).then((result) => ports.filter((_, idx) => result[idx]))
```
Comment (guzhongren) [inline frontend/scripts/runE2eWithServer.ts:24]: should not use `await` and `then` at the same time

File: frontend/scripts/runE2eWithServer.ts [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+    const availablePorts = ports.filter((port) => !usingPorts.includes(port))
+
+    console.log(`E2E: Currently available ports: ${availablePorts.join(',')}`)
```
Comment (guzhongren) [inline frontend/scripts/runE2eWithServer.ts:27]: - the prefix `E2E:` is not necessary
- follow our log convention, update to `Successfully to do ....` at least

File: frontend/scripts/runE2eWithServer.ts [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+      stderr: 'inherit',
+      shell: true,
+    })`./gradlew bootRun --args='--spring.profiles.active=local --MOCK_SERVER_URL=http://localhost:4323'`
+
+  const startSTUB = () =>
+    $({
+      cwd: DIR.STUB,
+      stderr: 'inherit',
+      shell: true,
+    })`docker-compose up`
```
Comment (guzhongren) [inline frontend/scripts/runE2eWithServer.ts:49]: should not use `multiple tech`
Maybe you can wrap all kinds of services into 1 `docker-compose.yaml` and then start it. ref to [this](https://stackoverflow.com/questions/53649873/docker-compose-mapping-local-directory-to-dockerfile-volume)

File: frontend/scripts/runE2eWithServer.ts [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+
+  const startServices = async (availablePorts: number[]) => {
+    if (availablePorts.length <= 0) {
+      return []
+    }
+
+    const processes = availablePorts.map((port) => {
+      switch (port) {
+        case PORT.BE:
+          console.log(`E2E: Try to start BE service`)
```
Comment (guzhongren) [inline frontend/scripts/runE2eWithServer.ts:59]: refactor log

File: frontend/scripts/runE2eWithServer.ts [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+          console.log(`E2E: Try to start STUB service`)
+          return startSTUB()
+      }
+    })
+
+    await Promise.all(
+      availablePorts.map((port) =>
+        tcpPortUsed
+          .waitUntilUsed(port, 1000, 30000)
+          .then(() => console.log(`E2E: Port ${port} is now started`))
```
Comment (guzhongren) [inline frontend/scripts/runE2eWithServer.ts:74]: await then

File: frontend/scripts/runE2eWithServer.ts [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+    newlyStartProcesses.forEach((pro) => {
+      console.log(`E2E: clean up process ${pro?.pid}`)
+      pro?.kill()
+    })
+
+    process.exit()
+  })
+}
+
+const [execPath, scriptPath, ...args] = process.argv
```
Comment (guzhongren) [inline frontend/scripts/runE2eWithServer.ts:103]: `execPath, scriptPath`  are assigned a value but never used

**Review by guzhongren — COMMENTED** [review]
Change with the comments

**Review by Zxx-Ivy — APPROVED** [review]
Done

**Pull request merged**